chain/stmgr/execute.go

```diff
@@ -128,10 +128,38 @@ func (sm *StateManager) ExecutionTraceWithMonitor(ctx context.Context, ts *types
 }
 
 func (sm *StateManager) ExecutionTrace(ctx context.Context, ts *types.TipSet) (cid.Cid, []*api.InvocResult, error) {
+	tsKey := ts.Key()
+
+	{
+		// check if we have the trace for this tipset in the cache
+		sm.execTraceCacheLock.Lock()
```

Comment:
Do we need a lock? I thought the ARCCache is thread-safe.

Reply:
For example, if two threads (T1,T2) enter this function with the same TipSet key and its not in the cache. Let say T1 finishes and stores in the cache. Then T3 enters with the same TipSet key as T1 and T2. For T3 it gets a cache hit. While its making the deep copy, then lets say T2 finishes and updates the cache. This means that when making the deepcopy on a cache hit other threads may update its value. This is therefore a critical section and needs to be guarded.

Reply:
Yeah, I _think_ you're right, and obviously better safe than sorry.

I am, however, concerned about slowing down this workflow by effectively forcing only one ExecutionTrace to be computed (or loaded-and-copied) at a time.

What do we think about dropping the lock before beginning an _actual_ computation of the ExecutionTrace (so line 142), making the deep copy, and then retaking it before writing it to the cache.

Reply:
@arajasek Ah, I thought the scope I introduced in  (in line 133-141) would run the defer when it exits the scope, however that is probably not the case and how defer works in Go (it only does that on function exit). Good catch!